sources/BigInt.swift

```diff
@@ -259,6 +259,17 @@ extension BigInt: IntegerArithmetic {
     public static func %(a: BigInt, b: BigInt) -> BigInt {
         return BigInt(abs: a.abs % b.abs, negative: a.negative)
     }
+  
+    /// Return the result of `a` mod `b`.
+    public static func modulus(a: BigInt, b: BigInt) -> BigInt {
+        if a.negative == b.negative {
+            return BigInt(abs: a.abs % b.abs, negative: a.negative)
+        }
+        else {
+            let floor = ((a / b) - 1) * b
+            return BigInt(abs: (floor.abs - a.abs) % b.abs, negative: b.negative)
```

Comment:
This seems suboptimal to me; can't we calculate the modulus using a single `%` operation on the absolute values? `a.abs % b.abs` should already get us close to the value we need; we just sometimes need to subtract the result from `b` to flip it.

Also I suspect there is a problem with the sign of the result. `a % b` must have the same sign as `a`, but I believe we wanted `BigInt.modulus(a, b)` to always return a nonnegative value. (Like the function [mpz_mod in GMP][gmp].)

[gmp]: https://gmplib.org/manual/Integer-Division.html

Something like this code may work:

```swift
let remainder = a.abs % b.abs
return a.negative && !remainder.isEmpty ? b.abs - remainder : remainder
```